### PR TITLE
update:fix headers and add progress function for hlsdl

### PR DESCRIPTION
**Pull request by yukunwu**
1.add custom headers for m3u8 url,add output name
2.use resty as http client with retry
3.expose progress
4.upgrade go module

### CODE REVIEW

File: hlsdl.go [hunk 1]
```diff
@@ -72,96 +76,63 @@ func wait(wg *sync.WaitGroup) chan bool {
...
-	}
-	defer file.Close()
-
-	if _, err := io.Copy(file, res.Body); err != nil {
-		return err
-	}
-
-	return nil
+	hlsDl.client.SetRetryCount(5).SetRetryWaitTime(time.Second)
+	_, err := hlsDl.client.R().SetHeaders(hlsDl.headers).SetOutput(segment.Path).Get(segment.URI)
```
Comment (canhlinh) [inline hlsdl.go:80]: @yukunwu Does Resty check the httpCode here?

Comment (yukunwu) [reply hlsdl.go:80]: No,I will fix it.

File: hlsdl.go [hunk 2]
```diff
@@ -179,67 +157,121 @@ func (hlsDl *HlsDl) downloadSegments(segments []*Segment) error {
...
+		return nil, nil, errors.New("Failed to get descryption key")
+	}
+	key = res.Body()
+	iv = []byte(segment.Key.IV)
+	if len(iv) == 0 {
+		iv = defaultIV(segment.SeqId)
+	}
+	return
+}
+func (hlsDl *HlsDl) GetProgress() float64 {
```
Comment (canhlinh) [inline hlsdl.go:269]: @yukunwu 

Showing progress is usually a real-time task. I think this function seems not to be fashionable. I suggest opening a channel that allows others to subscribe to listen to the progress's changes. What do you think?

Comment (yukunwu) [reply hlsdl.go:269]: I agree with you.I will remove this function.We can add this in next version.

**Pull request merged**